src/thread/queue.cr

```diff
@@ -0,0 +1,53 @@
+# Based on rubysl-thread ruby gem implementation
+# Copyright (c) 2013, Brian Shirai
+# Licensed under the BSD 3-clause
+
+# :nodoc:
+class Thread(T, R)
+  # :nodoc:
+  class Queue(T)
+    class Error < Exception; end
+
+    def initialize
+      @que = Deque(T).new(16)
+      @mutex = Thread::Mutex.new
```

Comment:
Using a mutex is easy, but maybe not optimal in terms of speed. Maybe using http://llvm.org/releases/3.3/docs/LangRef.html#cmpxchg-instruction and http://llvm.org/releases/3.3/docs/LangRef.html#atomicrmw-instruction for acquiring/releasing a lock could lead to better performance.